src/mpid/ch4/src/ch4_impl.h

```diff
@@ -396,18 +396,48 @@ static inline int MPIDIU_valid_group_rank(MPIR_Comm * comm, int rank, MPIR_Group
  * ALLFUNC_MUTEX lock forever is to insert YIELD in each loop. We choose to
  * insert it here for simplicity, but this might not be the best place. One
  * needs to investigate the appropriate place to yield the lock. */
+/* NOTE: Taking off VCI lock is necessary to avoid recursive locking and allow
+ * more granular per-vci locks */
+/* TODO: MPIDI_global.vci_lock probably will be changed into granular generic lock
+ */
 
 #define MPIDIU_PROGRESS()                                   \
     do {                                                        \
+        MPID_THREAD_CS_EXIT(VCI, MPIDI_global.vci_lock); \
         mpi_errno = MPID_Progress_test();                       \
+        MPID_THREAD_CS_ENTER(VCI, MPIDI_global.vci_lock); \
         MPIR_ERR_CHECK(mpi_errno);  \
         MPID_THREAD_CS_YIELD(GLOBAL, MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX); \
     } while (0)
 
+/* Optimized versions to avoid exessive locking/unlocking */
+/* FIXME: use inline function rather macros for cleaner semantics */
+
 #define MPIDIU_PROGRESS_WHILE(cond)         \
```

Comment:
Can we use inline function here instead of leaving a FIXME comment?

Reply:
I am trying to limit the changes in this PR. While it appears a lot of code change, it is still within the similar code design as original code. The `FIXME` is a reminder for further cleanup. `inline function` is just one idea. After the multi-VCI development, different solution may become better choice. For example, we may have better refactoring so the way RMA code calls progress may look different by then. That's why I would just leave as comment now and re-evaluate once the threading code settles.